src/v/datalake/translation/partition_translator.h

```diff
@@ -0,0 +1,166 @@
+    struct backoff {
+        static constexpr std::chrono::milliseconds initial_sleep{300};
+        explicit backoff(std::chrono::milliseconds max)
+          : max_sleep(max) {}
+
+        std::chrono::milliseconds next() {
+            current_sleep = std::min(
+              std::max(initial_sleep, current_sleep * 2), max_sleep);
+            return current_sleep;
+        }
+
+        void reset() { current_sleep = std::chrono::milliseconds{1}; }
+
+        template<
+          typename Func,
+          typename ShouldRetry,
+          typename FuncRet = std::invoke_result_t<Func>>
+        requires std::
+          predicate<ShouldRetry, typename ss::futurize<FuncRet>::value_type>
+          ss::futurize_t<FuncRet>
+          retry(int num_attempts, Func&& f, ShouldRetry&& should_retry) {
+            reset();
+            while (num_attempts-- > 0) {
+                auto result = co_await ss::futurize_invoke(f);
+                if (num_attempts > 0 && should_retry(result)) {
+                    co_await ss::sleep(next());
+                    continue;
+                }
+                co_return result;
+            }
+            __builtin_unreachable();
+        }
+
+        std::chrono::milliseconds current_sleep{1};
+        std::chrono::milliseconds max_sleep;
+    };
+
+    bool can_continue() const;
+
+    ss::future<std::optional<kafka::offset>> reconcile_with_coordinator();
+
+    ss::future<> do_translate();
+
+    using translation_success = ss::bool_class<struct translation_success>;
+    ss::future<translation_success> do_translate_once();
+    ss::future<model::record_batch_reader> make_reader();
+    ss::future<std::optional<coordinator::translated_offset_range>>
+      do_translation_for_range(kafka::read_committed_reader);
+
+    using checkpoint_result = ss::bool_class<struct checkpoint_result>;
+    ss::future<checkpoint_result>
+      checkpoint_translated_data(coordinator::translated_offset_range);
+
+    kafka::offset min_offset_for_translation() const;
+    // Returns max consumable offset for translation.
+    kafka::offset max_offset_for_translation() const;
+
+    model::term_id _term;
+    ss::lw_shared_ptr<cluster::partition> _partition;
+    ss::shared_ptr<translation_stm> _stm;
+    ss::sharded<coordinator::frontend>* _frontend;
+    ss::sharded<features::feature_table>* _features;
+    using jitter_t
+      = simple_time_jitter<ss::lowres_clock, std::chrono::milliseconds>;
+    jitter_t _jitter;
+    // Maximum number of bytes read in one go of translation.
+    // Memory usage tracking is not super sophisticated here, so we assume
+    // all data batches from the reader are buffered in the writer until
+    // they are flushed to disk. This is also factored into determining
+    // how many parallel translations can run at one point as we operate under
+    // a memory budget for all translations (semaphore below).
+    size_t _max_bytes_per_reader;
+    ssx::semaphore& _parallel_translations;
```

Comment:
consider using a pointer for consistency ?

Reply:
also if this is global per shard maybe we can wrap sempahore in a sharded object, this way it will be easier to add observability in future.